### DISCUSSION TITLE
()-indexing must appear last in an index expression?

### DISCUSSION THREAD

**Discussion by wendy444444** [yalmip/YALMIP#1119] [General discussion]
The premise is that Er and El are all binvar variables, and I have constrained
 `Implies (ER(I, T), El (J, T) = = 1)`
However, because my needs for different ER, I may constrain the same El, so I will report an error when I want to make EL equal to 0 when it is not equal to 1. What should I do? Looking at other people's examples, I know that they have added an index function, but I'm not sure that it can be done in yalmip, and if so, how can I define this index function? Thank you for your help.

**Comment by johanlofberg** [comment]
We don't know what error you are talking about ("I will report an error"??). 

I will make EL equal to 0 when it is not 1? If EL is a binvar, then trivially it will be 0 if it is not 1

As the code is now, you are using implies to code the trivial condition ER(I,T)==El(J,T)

**Reply by wendy444444** [reply]
Sorry, this error is an error reported by the running result of the program.
Because my program only constrains that when ER is equal to 1, the corresponding El is equal to 1, but it doesn't constrain when El is equal to 0, so in the process of optimization, the program optimizes the part of El that I need to be 0 to 1.
Actually, El represents the path of Er. If Er is 1, the El required by Er will be 1. However, there may be one Er with 0 and the other with 1, but some of their paths are the same, that is, the El of this same part is 1, so I can't simply make ER(I,T)==El(J,T).
Can you give me some advice, please?

**Reply by johanlofberg** [reply]
The error you have in the title is completely unrelated to YALMIP and is just something that happens when your MATLAB code is flawed

```
clear Y
Y{1}(1:2) = 1
clear Y
Y(1)(1:2) = 1
```

**Reply by wendy444444** [reply]
Thank you for your answer, sir. Excuse me, is it possible to put such a statement in constraints?

**Reply by johanlofberg** [reply]
What statement? I just showed you that the error pops up when you use non-valid MATLAB syntax

**Reply by wendy444444** [reply]
Ok, I thought this expression could be used to solve my problem.

**Reply by wendy444444** [reply]
> 

Sorry to bother you again, sir.
 I used the find function to constrain, 
`C = [C,(El(find(El(:,t)~=1),t)==0):'EL~=0'];`
but the program reported an error.
"Wrong find.
The function' find' corresponding to the input parameter of type' constraint' is not defined."

**Reply by johanlofberg** [reply]
Find is not supported.

But what on earth are you trying to do? That code simply says that if an element isn't 1, then it should be 0. That trivially holds if it is a binvar as that is the definition of being binary.

**Reply by wendy444444** [reply]
In the previous program, I constrained the case that El is equal to 1, but I didn't constrain the case that it is equal to 0. Although the program can run, from the results, it is equal to 1 when I need El to be equal to 0, so I want to make El that is not equal to 1 equal to 0, instead of optimizing it to 1 according to the result of the objective function.

**Reply by johanlofberg** [reply]
You are still very confusing. You are talking about a variable called El and you say "so I want to make El that is not equal to 1 equal to 0". What you are saying is that El is binary, i.e. you define at as an binvar and you are done.

**Reply by johanlofberg** [reply]
You have to define better what the problem is. El has both 0 and 1 in the optimal solution

```
>> value(El)

ans =

  Columns 1 through 8

     1     1     1     1     1     1     1     1
     1     1     1     1     1     1     1     1
     1     1     1     1     1     1     1     1
     1     1     1     1     1     1     1     1
     1     1     1     1     1     1     1     1
     0     0     1     1     1     1     1     1
     0     0     1     1     1     1     1     1
     0     0     1     1     1     1     1     1
     1     1     1     1     1     1     1     1
     0     0     1     1     1     1     1     1
     0     0     1     1     1     1     1     1

  Columns 9 through 10

     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
     1     1
```

**Reply by wendy444444** [reply]
Yes Sir, er is similar, I want el =1 when er=1, for the first column, er (6,1)=1, so I only need el(1, 1) and el(5,1)=1, but el(1:5, 1)=1 and el(2:4, 1)=1 is not needed, while el(2:4, 2)=1is what I want, so I think EL is 1 column ahead, because I'm not constraining  to el= 0, but I don't know how to constrain it right now. could you please help me?

**Reply by johanlofberg** [reply]
I don't see anything in the code that enforces El(6,1) to be 1 when Er(6,1) is 1. The closest you have is 

`implies(Er(i,t),El(Ek(i,p),t)==1)`

which for i=6,t=1 evaluates to 

`implies(Er(6,1),El([5 1],1)==1)`

which holds true
```

>> value(Er(6,1))

ans =

     1

>> value(El([5 1],1))

ans =

     1
     1

```

**Reply by wendy444444** [reply]
Yes, sir, the sentence you found is that I use Er to constrain El, but the values in this loop keep changing, and I can't write them all out. Moreover, the value of Er is determined by the solution of the objective function, so I don't know the value of ER until the solution is obtained.
Ek is the serial number of the required El corresponding to Er, and EL(6,1) has nothing to do with ER (6,1). I mean that when ER (6,1) = 1, EL(1,1)=1 and EL(5,1)=1,；when ER (5,2) = 1, El (1: 4，2）=1.
In the result, EL(1,1) and EL(5,1)=1 are due to ER (6,1) = 1. At this time, El (2: 4,1) should be equal to 0, but El (2: 4,1) = 1, which is earlier than I want El (1: 4,2) = 1.
(I made El(i,t-1)<=El(i,t) in the previous program, so once El(i, t)=1, then El(i, t: end) = 1)

**Reply by johanlofberg** [reply]
Nothing in your code says that El (2: 4,1) should be 0 in some particular case, nor are you penalizing those variables to guide the solver to never set a variable to 1 unless necessary. The solver cannot know what you want.

**Reply by wendy444444** [reply]
Yes, so I want to add a constraint. I tried the
 find function, 
El{:,t}{El{:,t}~=1}==0,
 implies (er (i, t)&j ~ = ek (i, p), el (j, t) = = 0),
 but all the programs reported errors. I don't know how to constrain it.

**Reply by johanlofberg** [reply]
As I said

`El{:,t}{El{:,t}~=1}==0`

simply states the obvious, El is binary.

Sounds like you simply want to constrain some elements to 1, and the other ones to 0.

```
these = Ek(i,p);
therest = setdiff(size(El,1),these);
...
implies(Er(i,t),El(these,t)==1)), implies(Er(i,t),El(therest,t)==0))
```

**Reply by wendy444444** [reply]
> Nothing in your code says that El (2: 4,1) should be 0 in some particular case, nor are you penalizing those variables to guide the solver to never set a variable to 1 unless necessary. The solver cannot know what you want.

I tried many times, including using the implied statement, 
`implied (er (i, t)&j ~ = ek (er (i, t) = = 1,1: ee (er (i, t) = = 1)), El (j, t) = = 0),`
 but there was still no way.Could you please help me again?

**Reply by wendy444444** [reply]
> 
Thank you for your kind help. I added the program statement according to your prompt, but it reported an error, and this result appeared.
![mmexport1652705218416](https://user-images.githubusercontent.com/103562806/168595673-b4cb49b3-7238-4078-8212-f78ef6bd89f5.png)

**Reply by johanlofberg** [reply]
Should have been

`these = Ek(i,1:EE(i));`

but then you also have to check so you don't add anything when threst is empty

**Reply by johanlofberg** [reply]
and you have to code around what happens when EE(i) is zero, since then I guess you have some empty case too on these

and it should be `setdiff(1:size(El,1),these)` of course

**Reply by wendy444444** [reply]
I tried again, and the result was Model is infeasible or unbounded. I think I need to add that different ERs may have overlapping Els, that is, different ERs have different constraints on the same El. Therefore, I think it will conflict to use EL (there, t) = = 0 directly.
(I think when EE(i)=0, p=1:EE(i) this loop will not go on because EE(i)=0<1)

**Reply by johanlofberg** [reply]
Code makes no sense now as you loop over p = 1:EE(i), and use a vectorized operation based on 1:EE(i)

The stuff you have talked about is simply

```
        these = Ek(i,1:EE(i));
        therest = setdiff(1:size(El,1),these);        
        if ~isempty(these)
            C = [C, implies(Er(i,t),El(these,t)==1)];
        end
        if ~isempty(therest)
            C = [C, implies(Er(i,t),El(therest,t)==0)];
        end
```

If you get some inconsistency, it means you should preprocess the data to avoid that

**Reply by wendy444444** [reply]
Thank you for your reply, I would like to ask whether I have dealt with the matter that different ER may have the same EL correctly. I am not sure, because the result of the program is wrong.
 I also tried the previous method in the remarks of the following line of therest, this line of program can run before, but after adding these and therest, it can not run, I also don't understand why, could you answer my question?

**Reply by johanlofberg** [reply]
As I said, you have to pre-process the data

Instead of
```
a = [1 2];
b = setdiff(1:n,a)
implies(x,y(a)==1) + implies(x,y(b)==0)
a = [5 6];
b = setdiff(1:n,a)
implies(x,y(a)==1) + implies(x,y(b)==0)
```

you have to do
```
a = [1 2];
a = [a 5 6]

b = setdiff(1:n,a)
implies(x,y(a)==1) + implies(x,y(b)==0)
```

**Reply by wendy444444** [reply]
I understand, thank you very much for your help. I'm sorry that I have to go to bed now. I will run our discussion as soon as I have time tomorrow. Wish you a happy life!

**Reply by wendy444444** [reply]
> As I said, you have to pre-process the data
> 
> Instead of
> 
> ```
> a = [1 2];
> b = setdiff(1:n,a)
> implies(x,y(a)==1) + implies(x,y(b)==0)
> a = [5 6];
> b = setdiff(1:n,a)
> implies(x,y(a)==1) + implies(x,y(b)==0)
> ```
> 
> you have to do
> 
> ```
> a = [1 2];
> a = [a 5 6]
> 
> b = setdiff(1:n,a)
> implies(x,y(a)==1) + implies(x,y(b)==0)
> ```

Hello, sir, after I preprocessed the data, I modified the program, but I got the result of Model is infeasible. I searched for a long time and only found that C [563: 585 587] is also infeasible, but I don't know how to deal with it, because in my opinion, the connection between them is not so close. What should I do? This is my program.
```

```

**Reply by johanlofberg** [reply]
https://yalmip.github.io/debugginginfeasible

**Reply by wendy444444** [reply]
I used the seventh one, and then I found that the same constraint is in a loop, and some will have problems, but some won't.

**Reply by wendy444444** [reply]
> Code makes no sense now as you loop over p = 1:EE(i), and use a vectorized operation based on 1:EE(i)
> 
> The stuff you have talked about is simply
> 
> ```
>         these = Ek(i,1:EE(i));
>         therest = setdiff(1:size(El,1),these);        
>         if ~isempty(these)
>             C = [C, implies(Er(i,t),El(these,t)==1)];
>         end
>         if ~isempty(therest)
>             C = [C, implies(Er(i,t),El(therest,t)==0)];
>         end
> ```
> 
> If you get some inconsistency, it means you should preprocess the data to avoid that

Hello, sir, I found that because I defined ER as sdpvar, the result of these and therest was Nan, so I defined them as sdpvar and added C = [C, more constraints] to them. However, if we do this, unique cannot be used, and the program reported an error: unique currently only supported for vectors.

**Reply by johanlofberg** [reply]
More things that don't make sense. "I defined ER as sdpvar, the result of these and therest was Nan, so I defined them as sdpvar"

unique is not applicable to sdpvars. From the general stuff you shown so far, there is absolutely no reason that you should have to use convoluted code using stuff as find/if/unique etc (which isn't supported anyway). Everything has to be written either using linear inequalities, or by implies operators involving sdpvars and constraints built from sdpvars (which yalmip converts to linear inequalitites).

**Reply by wendy444444** [reply]
> More things that don't make sense. "I defined ER as sdpvar, the result of these and therest was Nan, so I defined them as sdpvar"
> 
> unique is not applicable to sdpvars. From the general stuff you shown so far, there is absolutely no reason that you should have to use convoluted code using stuff as find/if/unique etc (which isn't supported anyway). Everything has to be written either using linear inequalities, or by implies operators involving sdpvars and constraints built from sdpvars (which yalmip converts to linear inequalitites).

Thank you for your reply. It's true that I want to use sdpvar type, but I don't know how to express my needs if I can't use unique. If you want to add intermediate variables for assistance, what format should you use?

**Reply by johanlofberg** [reply]
then you define more variables and connect the variables via constraints to represent the logic
https://yalmip.github.io/tutorial/logicprogramming
https://yalmip.github.io/example/modellingif/

**Reply by wendy444444** [reply]
Thank you for your help. When using the union function, the program reported the following error. Excuse me, is there a unique function in union?
![微信图片_20220522155926](https://user-images.githubusercontent.com/103562806/169684913-4616d9cb-3464-4918-b2f0-b55ba1ed462e.png)

**Reply by johanlofberg** [reply]
I thought I rather clearly told you that you cannot use stuff like that. You have to be able to represent it using *linear* constraints and operators, and implies, typically invoking temporary variables to represent more complex operations.

Despite the fact that the operator isn't supported, the thing you tell it to do makes absolutely no sense. You want to create a union (which would yield an expression with dimension 1 and up) and then insert that into a matrix with specified dimension.

**Reply by wendy444444** [reply]
Sir, but when I use implies, this error appears“Indexing logic not supported yet”. How can I express it?
` C = [C,(implies(i==1,these(i,EE(i)+1:end)==1)):'delta(i,1)'];`

**Reply by johanlofberg** [reply]
I assume i is a double, hence all you are saying is
```
if i == 1
  C = [C,these(i,EE(i)+1:end) ==1 ];
end
```
it is not a logic constraint, it is simple linear constraint as the logic is fixed and does not depend on any decision variable.

**Reply by johanlofberg** [reply]
assuming EE is constant. If EE is a decision variable, you will have do much much more advanced modelling

**Reply by wendy444444** [reply]
Sir, you're right. EE is a constant matrix. I tried your program, and it didn't report an error, but it was useless. The definition of this is sdpvar, which is still nan after the program runs.

**Reply by wendy444444** [reply]
> assuming EE is constant. If EE is a decision variable, you will have do much much more advanced modelling

Sir, I want to express whether i exists in the matrix these. The program using ismember is as follows
`for i = 1:11
C = [ C, (implies(~ismember(i,these( :, :, t )),ret11(i,t)==i )):'El(Ek(i,1:EE(i)),t)==1' ];
end` 
The program reported the error "For Colon Operator with Char Operands, First and Last Operands Must Be Char." So I guess ismember doesn't support it either, so how should I express it?

**Reply by johanlofberg** [reply]
To me that error is telling me these is a char since the only place a colon is is inside these, hence why on earth are you involving a fixed known data inside an implies. Once again, you don't seem to understand when you´re simply adding a trivial linear constraint vs modelling a combinatorial thing

**Reply by johanlofberg** [reply]
and if you have nans, it's either because you never solved the problem, or you are destroying the definition of EE somehow https://yalmip.github.io/naninmodel

**Reply by wendy444444** [reply]
> To me that error is telling me these is a char since the only place a colon is is inside these, hence why on earth are you involving a fixed known data inside an implies. Once again, you don't seem to understand when you´re simply adding a trivial linear constraint vs modelling a combinatorial thing

I'm sorry, sir, I didn't make myself clear, but these is a matrix of sdpvar type.That's why I use implies, but I don't know how to express what I want.

**Reply by johanlofberg** [reply]
Something is going wrong with these(:,:,t) so show what these is, what t is, and what happens when you type these(:,:,t). 

Regardless, you cannot use ismember, but have to encode it by introducing temporary variables which encode which values you have in that matrix, and then activate a second binary if wanted value is in there, and then use that negated activation variable in an implies. It would be a complicated several-layer deep model starting with a permutation matrix to model every element of the vector. 

Roughly speaking, y = ismember(i,x) for scalar integer decision variable x between 1 to N can be written as

```
where = binvar(1,N)
binvar y
[x == where*(1:N)', y == where(i)]
```

alternatively, you introduce binaries y, y1 and y2 to represent x == i, x>=i+1 and x<= i-1 respectively, and then add disjunction y+y1+y2==1

**Reply by wendy444444** [reply]
My definition of these is as follows.
```
numbus=12；tt=10；
these=sdpvar( numbus,5, tt, 'full' );
these( 1, 1, : ) = 0;
Er = binvar( numbus, tt, 'full' );
for t = 2:tt
    for i = 2:numbus
           C = [ C, (these( i, 1:EE( i ), t ) == Er( i, t )*Ek( i, 1:EE( i ) )):'these( i, 1:EE( i ), t )' ];
    end
end
```
The values of EE and Ek are in the picture.
![dba0bf875f5ec3aad72c00956abcbf8](https://user-images.githubusercontent.com/103562806/169694315-e1051a09-8f76-40c5-8d19-db2e2cd4b229.png)
![image](https://user-images.githubusercontent.com/103562806/169694337-3b1e1823-91c5-4a07-8d97-a4a48c1fa45c.png)

**Reply by johanlofberg** [reply]
and what happens when you write these(:,:,t) at the point where it crashes. you have to show what crashes and what things look like when you get the crash. It is basic debugging principles

**Reply by wendy444444** [reply]
```numbr=11；
ret11=sdpvar( numbr, tt, 'full' );
for i = 1:11 C = [ C, (implies(~ismember(i,these( :, :, t )),ret11(i,t)==i )):'El(Ek(i,1:EE(i)),t)==1' ]; end```

I used ismember, so the program reported the error“ for colon operator with char operands, first and last operands must be char.”

**Reply by johanlofberg** [reply]
Yes, but exactly what is causing the crash. The crash is due to some incorrect use of colon, or a bug in colon. In that pice of code, you use colon in the indexing of these, and you use a colon when you tag a constraint. Hence, which is it, and why does it happen (i.e. what is involved in the use of the colon). It could be that these is strange (thus look at it, and what is returned whren you do that indexing) or that the implies operator returns crap (I presume so, but you have to look at the result to confirm) causing that colon in the tagging to puke. You cannot just list the full code in which the error hides

**Reply by wendy444444** [reply]
```
         if  ~ismember(i,these( :, :, t ))
             C = [ C, ret11(i,t)==i ];
         end

```
OK, I rephrased this sentence. I think it's strange that the error is not in the line ismemer, but in the line C = [C,], so I deleted the comment to get it. However，when running again, yalmip becomes particularly slow, and twenty minutes have passed without any action.

**Reply by johanlofberg** [reply]
On closer inspection, ismember is overloaded on sdpvars, but it is not doing what you are tryin to use it for. It checks if the first element is structurally in the second, not whether the decision variables have the same values in the solution

```
>> x = sdpvar(1,2);
>> ismember(x(2),x)

ans =

  2×1 logical array

   0
   1
```

hence in your case it will always be false.

Your use of tagging fails as you've missed that : has precedence over logic equality

```
>> C = [];C=[C, x==1:'what']
Error using  : 
For colon operator with char operands, first and last operands
must be char.
>> C = [];C=[C, (x==1):'what']
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|   ID|                Constraint|   Coefficient range|    Tag|
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|   #1|   Equality constraint 1x2|              1 to 1|   what|
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

```

**Reply by wendy444444** [reply]
Sir, you're right. Such a mistake did happen.
